src/x/server/server.go

```diff
@@ -140,11 +156,78 @@ func (s *server) Serve(l net.Listener) error {
 	return nil
 }
 
+func (s *server) isTLSConfigCacheValid() bool {
+	if s.tlsConfigCache == nil || s.certsTTLTicker == nil {
+		return false
+	}
+	select {
+	case <-s.certsTTLTicker.C:
+		return false
+	default:
+		return true
+	}
+}
+
+func (s *server) getTLSConfig() (*tls.Config, error) {
+	s.Lock()
+	defer s.Unlock()
+	if s.isTLSConfigCacheValid() {
+		return s.tlsConfigCache, nil
+	}
+	if s.tlsOpts.ClientCAFile() != "" {
+		certs, err := os.ReadFile(s.tlsOpts.ClientCAFile())
+		if err != nil {
+			return nil, fmt.Errorf("read bundle error: %w", err)
+		}
+		if ok := s.certPool.AppendCertsFromPEM(certs); !ok {
+			return nil, fmt.Errorf("cannot append cert to cert pool")
+		}
+	}
+	clientAuthType := tls.NoClientCert
+	if s.tlsOpts.MutualTLSEnabled() {
+		clientAuthType = tls.RequireAndVerifyClientCert
+	}
+	cert, err := tls.LoadX509KeyPair(s.tlsOpts.CertFile(), s.tlsOpts.KeyFile())
+	if err != nil {
```

Comment:
Say we already have a currently valid TLS config, but load of the new TLS config loads. Will we immediately start erroring our connections? I'm wondering whether we should degrade more gracefully than that:

- Try to load new TLS config
- If that fails: keep using the old one (and alert)

Would that be reasonable?

Reply:
Yes, it sounds reasonable. I fixed it. For now, if it fails to read the new TLS config, it will increase an error metric and write an error message into the log.